libfoundation/src/foundation-foreign.cpp

```diff
@@ -182,545 +321,403 @@ bool __MCForeignValueCopyDescription(__MCForeignValue *self, MCStringRef& r_desc
+        r_boxed_value = MCValueRetain(p_value ? kMCTrue : kMCFalse);
+        return true;
     }
-#ifdef __64_BIT__
-    else if (t_value >= -(1LL << std::numeric_limits<double>::digits) &&
-             t_value <= (1LL << std::numeric_limits<double>::digits))
-    {
-        return MCNumberCreateWithReal(double(t_value),
-                                      reinterpret_cast<MCNumberRef&>(r_value));
-    }
-#endif
-    else
-    {
-        return __throw_numeric_overflow(kMCForeignImportErrorTypeInfo,
-                                        kMCSSizeTypeInfo);
-	}
-}
-
-static bool __float_import(void *contents, bool release, MCValueRef& r_value)
-{
-    return MCNumberCreateWithReal(*(float *)contents, (MCNumberRef&)r_value);
-}
+};
 
-static bool __double_import(void *contents, bool release, MCValueRef& r_value)
+template <>
+struct DoExport<bool_type_desc_t>
 {
-    return MCNumberCreateWithReal(*(double *)contents, (MCNumberRef&)r_value);
-}
-
-template <typename T>
-static bool
-__any_int_export(MCValueRef value,
-                 bool release,
-                 void *contents,
-                 MCTypeInfoRef typeinfo)
-{
-    // Fetch the number as a double
-    double t_value =
-            MCNumberFetchAsReal(static_cast<MCNumberRef>(value));
- 
-    // First check that the value is within the contiguous integer range
-    // of doubles. If that succeeds, then check it fits within the target
-    // integer type.
-    if (t_value < double(-(1LL << std::numeric_limits<double>::digits)) ||
-        t_value > double(1LL << std::numeric_limits<double>::digits) ||
-        t_value < double(std::numeric_limits<T>::min()) ||
-        t_value > double(std::numeric_limits<T>::max()))
+    static bool doexport(MCValueRef p_boxed_value, bool& r_value)
     {
-        return __throw_numeric_overflow(kMCForeignExportErrorTypeInfo,
-                                        typeinfo);
+        r_value = (p_boxed_value == kMCTrue);
+        return true;
     }
+};
 
-    *(T *)contents = T(t_value);
+/* ---------- pointer specializations */
 
-    if (release)
-        MCValueRelease(value);
-    
-    return true;
-}
+/* ---------- floating-point numeric specializations */
 
-static bool
-__int_export(MCValueRef value, bool release, void *contents)
+template <typename RealType>
+struct DoExport<
+    RealType, typename std::enable_if<std::is_floating_point<typename RealType::c_type>::value>::type>
 {
-    return __any_int_export<integer_t>(value, release, contents, kMCIntTypeInfo);
-}
-
-static bool
-__uint_export(MCValueRef value, bool release, void *contents)
-{
-	return __any_int_export<uinteger_t>(value, release, contents, kMCUIntTypeInfo);
-}
-
-static bool
-__size_export(MCValueRef value, bool release, void *contents)
-{
-    return __any_int_export<size_t>(value, release, contents, kMCSizeTypeInfo);
-}
+    static bool doexport(MCNumberRef p_boxed_value, typename RealType::c_type& r_value)
```

Comment:
It might be worth extracting the C type as a template parameter, just in order to avoid repeating `typename RealType::c_type`.  Don't forget we're allowed default template parameters now!

```c++
template <typename TypeDesc, typename RealType = typename TypeDesc::c_type>
struct DoExport<TypeDesc,
    typename std::enable_if<std::is_floating_point<RealType>::value>::type>
{
    static bool doexport(MCNumberRef p_boxed_value, RealType& r_value)
    {
        r_value = RealType(MCNumberFetchAsReal(p_boxed_value));
        return true;
    }
}
```

I guess you could apply this pattern in several places where you're currently tediously extracting the `c_type` from the type descriptor struct.